server/src/main/java/org/apache/iotdb/db/metadata/MTree.java

```diff
@@ -781,20 +777,56 @@ private void findStorageGroup(MNode node, String path, HashSet<String> res) {
    *
    * @return a list contains all distinct devices
    */
-  Set<String> getAllDevices() {
-    return new HashSet<>(getNodesList(3));
+  Set<String> getAllDevices() throws SQLException {
+    List<String> res = new ArrayList<>();
+    MNode node;
+    if ((node = getRoot()) != null) {
+      findDevices(node, "root", res);
+    }
+    return new LinkedHashSet<>(res);
+  }
+
+  private void findDevices(MNode node, String path, List<String> res) {
```

Comment:
This function is very confused, what's the relationship between node and path?

Reply:
@qiaojialin It might not have to declare two separate params here, which it could use node.getName() to get node's name. However, the function above (findStorageGroup) has also declared like this, so I'm not sure if I have to change this declaration.

Reply:
Or are you confusing with the naming of params? Is there any suggestion about how to change it to resolve the confusion?

Reply:
Here path != node.getName()  , path is the full path (e.g., root.d1.s1) of the node while node.getName only return the node name (e.g., s1). 

But now, after you change the res.add(path), these two parameters are consistent. No need to change.

BTW, if you find some existing codes which are confusing, do not hesitate to optimize them.